### PR TITLE
Accept gradient iterators in global norm helpers

### PR DESCRIPTION
**Pull request by vineethsaivs**
Global norm calculation consumes a gradient iterator during validation and then raises; clipping a generator cannot complete a training step.

**Cause:** The helper advertises Iterable input but traverses it repeatedly and assumes len and indexing. The clipping wrapper needs the tensors again after norm calculation.

**Fix:** Materialize the input iterable once in each public helper before it is reused.

**Test:** `DS_ACCELERATOR=cpu python -m pytest tests/unit/runtime/test_runtime_utils.py -k 'global_norm_accepts_iterables or clip_tensor_generator or test_call_to_str' -q`. Five iterator regressions fail before. Fourteen targeted tests pass after: list/tuple/iterator norms for p=1,2,3,infinity, two SGD steps compared with PyTorch clipping, and an existing utility control. Validated on Apple Silicon CPU with PyTorch 2.10.0. No GPU or distributed run. Changed-file pre-commit passes.

### CODE REVIEW

File: deepspeed/runtime/utils.py [hunk 1]
```diff
@@ -901,6 +901,7 @@ def get_global_norm_of_tensors(input_tensors, norm_type=2, mpu=None, use_graph=F
         Total norm of the tensors (viewed as a single vector).
     """
     assert isinstance(input_tensors, Iterable), f'expected Iterable type not {type(input_tensors)}'
+    input_tensors = list(input_tensors)
```
[the remark this answers is not in the document]
Comment (vineethsaivs) [reply deepspeed/runtime/utils.py:904]: The signoff is already present in `7127b87d`: `Signed-off-by: Vineeth Sai <vineethsai4444@gmail.com>`. All non-merge commits in this PR have it.
